File: Leetcode/Convert Sorted List to Binary Search Tree.cc
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode *sortedListToBST(ListNode *head) {
        if (!head) {
            return NULL;
        }
        return sortedListToBST(head, NULL);
    }
private:
    TreeNode *sortedListToBST(ListNode *begin, ListNode *end) {
        if (begin == end) {
            return NULL;
        }
        ListNode *mid = get_mid(begin, end);
        TreeNode *root = new TreeNode(mid -> val);
        root -> left = sortedListToBST(begin, mid);
        root -> right = sortedListToBST(mid -> next, end);
        return root;
    }
    
    ListNode *get_mid(ListNode *begin, ListNode *end) {
        if (begin == end) {
            return NULL;
        }
        ListNode *a, *b;
        a = b = begin;
        while (true) {
            if (b -> next == end) {
                break;
            }
            b = b -> next;

            if (b -> next == end) {
                break;
            }
            b = b -> next;
            
            if (a -> next == end) {
                break;
            }
            a = a -> next;
        }
        return a;
    }
};

File: Leetcode/python/Convert Sorted List to Binary Search Tree.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
#
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a list node
    # @return a tree node
    def sortedListToBST(self, head):
        n, cur = 0, head
        while cur:
            n += 1
            cur = cur.next
        return self.do_sortedListToBST(n, head)
    
    def do_sortedListToBST(self, n, head):
        if not n:
            return None
        prev, cur = None, head
        for i in xrange(n / 2):
            prev = cur
            cur = cur.next
        next = cur.next
        root = TreeNode(cur.val)
        if prev:
            prev.next = None
        root.left = self.do_sortedListToBST(n / 2, head)
        root.right = self.do_sortedListToBST(n - n / 2 - 1, next)
        return root
